File: py/crud.py
# import file dan library
from mahasiswa import Mahasiswa


class Crud:  # deklarasi class
    """ class crud berisi list of object dari class mahasiswa
    yang digunakan untuk memodifikasi vector tersebut, seperti
    menambahkan data, menampilkan data, memperbaharui data,
    dan menghapus data
    """
    # CONSTRUCTOR

    def __init__(self, data: list):
        """ konstruktor yang menerima parameter dengan tipe data list sebagai inputannya """
        self.__data = data

    # PUBLIC METHOD

    def tampilData(self):
        """ method untuk menampilkan data yang ada di dalam list """
        print("Data Mahasiswa : ")
        for x in self.__data:
            print("==========================================")
            print(x)

    def tambahData(self, nama, nim, prodi, fakultas):
        """ method untuk menambahkan data baru ke dalam list """
        temp = Mahasiswa(nama, nim, prodi, fakultas)  # instansiasi objek class
        self.__data.append(temp)  # menambahkannya ke list

    def cariNIM(self, nim):
        """ method yang digunakan untuk memudahkan pencarian NIM ketika
         ingin memperbaharui atau menghapus data """
        if nim not in Mahasiswa._allNIM:
            # jika nim yang dicari tidak ada di list NIM
            return -1
        else:
            # jika nim ditemukan maka akan mengembalikan indexnya
            return Mahasiswa._allNIM.index(nim)

    def ubahData(self, indexNIM, nama, nim, prodi, fakultas):
        """ method untuk memperbaharui data mahasiswa """
        # mengupdate data lama dengan data baru melalui index dengan setter
        self.__data[indexNIM].nama = nama
        self.__data[indexNIM].nim = nim
        self.__data[indexNIM].prodi = prodi
        self.__data[indexNIM].fakultas = fakultas
        Mahasiswa._allNIM[indexNIM] = nim

    def hapusData(self, indexNIM):
        """ method untuk menghapus data mahasiswa """
        # memanggil fungsi pop untuk mengeluarkan data yang ingin dihapus
        self.__data.pop(indexNIM)

File: php/Crud.php
<?php
// import file
require("Mahasiswa.php");

class Crud{ //deklarasi class
    /* class Crud berisi array of object dari class mahasiswa 
    yang digunakan untuk mmeodifikasi array tersebut, seperti
    menambahkan data, menampilkan data, memperbaharui data, dan
    menghapus data */

    // private atribut
    private $dataMhs = array();

    // CONSTRUCT
    public function __construct($data = array()){
        /* konstruktor yang menerima parameter dengan tipe data array sebagai inputannya*/ 
        $this->dataMhs = $data;
    }

    // PUBLIC METHOD
    public function tambahData($idx, $nama = "", $nim = "", $prodi = "", $fakultas = "", $foto = ""){
        /*method yang digunakan untuk menambahkan data baru
        ke dalam vector*/
        $this->dataMhs[$idx] = new Mahasiswa($nama, $nim, $prodi, $fakultas, $foto); #instansiasi objek mahasiswa
    }

    public function tampilData(){
        /*method yang digunakan untuk menampilkan data mahasiswa
        yang ada di dalam vector*/
        echo "<table border='1'>";
        echo "<tr>
            <th>Nama</th>
            <th>NIM</th>
            <th>Program Studi</th>
            <th>Fakultas</th>
            <th>Foto Profil</th>
            </tr>";
        for($i = 0; $i < sizeof($this->dataMhs); $i++){
            echo "<tr><td>";
            echo $this->dataMhs[$i]->getNama();
            echo "</td><td>";
            echo $this->dataMhs[$i]->getNIM();
            echo "</td><td>";
            echo $this->dataMhs[$i]->getProdi();
            echo "</td><td>";
            echo $this->dataMhs[$i]->getFakultas();
            echo "</td><td>";
            echo "<img src='" . $this->dataMhs[$i]->getFoto(). "'style='width:100px'>";
            echo "</td>";
            echo "</tr>";
        }
        echo "</table>";
    }

    function cariNIM($nim = ""){
        /*method yang digunakan untuk mencari nim
        dan mengembalikan index nim tersebut*/
        for($i = 0; $i < sizeof($this->dataMhs); $i++){
            if(strcmp($nim, $this->dataMhs[$i]->getNIM()) == 0){
                return $i; //mengembalikan index jika nim tersebut ditemukan
            }
        }
        return -1;
    }

    public function updateData($idx, $nama = "", $nim = "", $prodi = "", $fakultas = "", $foto = ""){
        /*method yang digunakan untuk memperbaharui data mahasiswa
        pada index tertentu*/
        // memanggil setter untuk setiap atribut
        $this->dataMhs[$idx]->setNama($nama);
        $this->dataMhs[$idx]->setNIM($nim);
        $this->dataMhs[$idx]->setProdi($prodi);
        $this->dataMhs[$idx]->setFakultas($fakultas);
        $this->dataMhs[$idx]->setFoto($foto);
    }

    public function hapusData($idx){
        /*method yang digunakan untuk menghapus data mahasiswa
        pada index tertentu*/
        array_splice($this->dataMhs, $idx, 1); # menggunakan array_splice untuk menghapus data dan reindexing
    } 
    
    // DESTRUCTOR
    function __destruct(){}
}

?>